File: modules/local-connection/android/src/main/java/expo/modules/localconnection/LocalConnectionModule.kt
package expo.modules.localconnection

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.nio.charset.StandardCharsets

class LocalConnectionModule : Module() {
  
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context lost")

  // Lazy load manager with event wiring
  private val sessionManager: NearbyConnectionManager by lazy {
    val manager = NearbyConnectionManager(context)
    
    // 1. Discovery: Found a room/peer
    manager.onOpponentFound = { endpointId, endpointName ->
      sendEvent("onChange", mapOf(
        "type" to "found", 
        "endpointId" to endpointId,
        "endpointName" to endpointName
      ))
    }
    
    manager.onEndpointLost = { endpointId ->
      sendEvent("onChange", mapOf(
        "type" to "lost", 
        "endpointId" to endpointId
      ))
    }

    // 2. Connection Request: Someone wants to join us
    manager.onConnectionRequest = { endpointId, endpointName ->
      sendEvent("onChange", mapOf(
        "type" to "invite",
        "endpointId" to endpointId,
        "endpointName" to endpointName
      ))
    }
    
    // 3. Connection Established
    manager.onOpponentConnected = { peerName ->
      sendEvent("onChange", mapOf(
        "type" to "connected", 
        "peer" to peerName
      ))
    }
    
    // 4. Connection Lost
    manager.onOpponentDisconnected = { peerName ->
      sendEvent("onChange", mapOf(
        "type" to "disconnected", 
        "peer" to peerName
      ))
    }
    
    // 5. Data Received
    manager.onDataReceived = { data, peerName ->
      val message = String(data, StandardCharsets.UTF_8)
      sendEvent("onChange", mapOf(
        "type" to "data", 
        "data" to message, 
        "peer" to peerName
      ))
    }
    
    manager // return the instance
  }

  override fun definition() = ModuleDefinition {
    Name("LocalConnection")
    Events("onChange")

    // --- SETUP ---

    Function("InitPeerName") { name: String ->
      sessionManager.setPeerName(name)
    }

    Function("getOpponentName") {
      sessionManager.getOpponentName()
    }

    // Set connection medium: "BLUETOOTH", "WIFI", or "ALL"
    Function("setConnectionMedium") { medium: String ->
      sessionManager.setMedium(medium)
    }

    Function("startAdvertising") {
      // NOTE: Ensure 'startAdvertising()' is public in NearbyConnectionManager
      // OR use sessionManager.start() if you want to trigger both at once.
      sessionManager.startAdvertising()
    }

    Function("startScanning") {
      // NOTE: Ensure 'startDiscovery()' is public in NearbyConnectionManager
      sessionManager.startDiscovery()
    }

    Function("stopAdvertising") {
      sessionManager.stopAdvertising()
    }

    Function("stopScanning") {
      sessionManager.stopDiscovery()
    }

    Function("EndConnection") {
      sessionManager.endConnection()
    }

    // --- CONNECTION HANDSHAKE (Crucial for P2P) ---

    // Called by JS when user clicks a "Found" room
    Function("joinRoom") { endpointId: String ->
      sessionManager.joinRoom(endpointId)
    }

    // Called by JS when user accepts an "Invite"
    Function("acceptInvitation") { endpointId: String ->
      sessionManager.acceptConnection(endpointId)
    }

    // --- GAMEPLAY ---

    AsyncFunction("sendData") { message: String ->
      val data = message.toByteArray(StandardCharsets.UTF_8)
      sessionManager.send(data)
    }
  }
}

File: modules/local-connection/ios/LocalConnectionModule.swift
import ExpoModulesCore

public class LocalConnectionModule: Module {
  private lazy var sessionManager: CoreBluetoothSessionManager = {
    let manager = CoreBluetoothSessionManager()
    
    // 1. Discovery: Found a room/peer
    manager.onOpponentFound = { [weak self] endpointId, endpointName in
      self?.sendEvent("onChange", [
        "type": "found",
        "endpointId": endpointId,
        "endpointName": endpointName
      ])
    }
    
    manager.onEndpointLost = { [weak self] endpointId in
      self?.sendEvent("onChange", [
        "type": "lost",
        "endpointId": endpointId
      ])
    }
    
    // 2. Connection Request: Someone wants to join us
    manager.onConnectionRequest = { [weak self] endpointId, endpointName in
      self?.sendEvent("onChange", [
        "type": "invite",
        "endpointId": endpointId,
        "endpointName": endpointName
      ])
    }
    
    // 3. Connection Established
    manager.onPeerConnected = { [weak self] peerName in
      self?.sendEvent("onChange", [
        "type": "connected",
        "peer": peerName
      ])
    }
    
    // 4. Connection Lost
    manager.onPeerDisconnected = { [weak self] peerName in
      self?.sendEvent("onChange", [
        "type": "disconnected",
        "peer": peerName
      ])
    }
    
    // 5. Data Received
    manager.onDataReceived = { [weak self] data, peerName in
      if let message = String(data: data, encoding: .utf8) {
        self?.sendEvent("onChange", [
          "type": "data",
          "data": message,
          "peer": peerName
        ])
      }
    }
    
    return manager
  }()

  public func definition() -> ModuleDefinition {
    Name("LocalConnection")
    Events("onChange")

    // --- SETUP ---

    Function("InitPeerName") { (name: String) in
      self.sessionManager.setPeerName(name)
    }

    Function("getOpponentName") {
      return self.sessionManager.getOpponentName()
    }

    // iOS MultipeerConnectivity doesn't support medium selection - it auto-negotiates
    // This is a no-op stub for API consistency with Android
    Function("setConnectionMedium") { (medium: String) in
      print("LocalConnection: setConnectionMedium('\(medium)') - iOS auto-selects transport")
    }

    Function("startAdvertising") {
      self.sessionManager.startAdvertising()
    }

    Function("startScanning") {
      self.sessionManager.startScanning()
    }

    Function("stopAdvertising") {
      self.sessionManager.stopAdvertising()
    }

    Function("stopScanning") {
      self.sessionManager.stopScanning()
    }

    Function("EndConnection") {
      self.sessionManager.endConnection()
    }

    // --- CONNECTION HANDSHAKE ---

    // Called by JS when user clicks a "Found" room
    Function("joinRoom") { (endpointId: String) in
      self.sessionManager.joinRoom(endpointId)
    }

    // Called by JS when user accepts an "Invite"
    Function("acceptInvitation") { (endpointId: String) in
      self.sessionManager.acceptInvitation(endpointId)
    }

    // --- GAMEPLAY ---

    AsyncFunction("sendData") { (message: String) in
      if let data = message.data(using: .utf8) {
        self.sessionManager.send(data: data)
      }
    }
  }
}
